src/vs/workbench/contrib/terminal/browser/voiceTerminalActions.ts

```diff
@@ -0,0 +1,211 @@
+	}
+}
+
+
+export class StopTerminalSpeechToTextAction extends Action2 {
+
+	static readonly ID = 'workbench.action.stopTerminalSpeechToText';
+
+	constructor() {
+		super({
+			id: 'workbench.action.stopTerminalSpeechToText',
+			title: {
+				value: localize('workbench.action.stopTerminalSpeechToText', "Stop Terminal Speech To Text"),
+				original: 'Stop Terminal Speech To Text'
+			},
+			precondition: ContextKeyExpr.and(HasSpeechProvider, TerminalContextKeys.focus),
+			f1: true
+		});
+	}
+
+	async run(accessor: ServicesAccessor): Promise<void> {
+		const instantiationService = accessor.get(IInstantiationService);
+		TerminalVoiceSession.getInstance(instantiationService).stop(true);
+	}
+}
+
+export class TerminalVoiceSession extends Disposable {
+	private _input: string = '';
+	private _decoration: IDecoration | undefined;
+	private _marker: IXtermMarker | undefined;
+	private static _instance: TerminalVoiceSession | undefined = undefined;
+	private _acceptTranscriptionScheduler: RunOnceScheduler | undefined;
+	static getInstance(instantiationService: IInstantiationService): TerminalVoiceSession {
+		if (!TerminalVoiceSession._instance) {
+			TerminalVoiceSession._instance = instantiationService.createInstance(TerminalVoiceSession);
+		}
+
+		return TerminalVoiceSession._instance;
+	}
+	private _cancellationTokenSource: CancellationTokenSource | undefined;
+	private _disposables = new DisposableStore();
+	constructor(
+		@ISpeechService private readonly _speechService: ISpeechService,
+		@ITerminalService readonly _terminalService: ITerminalService,
+		@IConfigurationService readonly configurationService: IConfigurationService,
+		@IInstantiationService readonly _instantationService: IInstantiationService
+	) {
+		super();
+		this._register(this._terminalService.onDidChangeActiveInstance(() => this.stop()));
+		this._register(this._terminalService.onDidDisposeInstance(() => this.stop()));
+	}
+
+	start(): void {
+		this.stop();
+		let voiceTimeout = this.configurationService.getValue<number>(AccessibilityVoiceSettingId.SpeechTimeout);
+		if (!isNumber(voiceTimeout) || voiceTimeout < 0) {
+			voiceTimeout = SpeechTimeoutDefault;
+		}
+		this._acceptTranscriptionScheduler = this._disposables.add(new RunOnceScheduler(() => {
+			this._terminalService.activeInstance?.sendText(this._input, false);
+			this.stop();
+		}, voiceTimeout));
+		this._cancellationTokenSource = new CancellationTokenSource();
+		const session = this._disposables.add(this._speechService.createSpeechToTextSession(this._cancellationTokenSource!.token));
+
+		this._disposables.add(session.onDidChange((e) => {
+			if (this._cancellationTokenSource?.token.isCancellationRequested) {
+				return;
+			}
+			switch (e.status) {
+				case SpeechToTextStatus.Started:
+					// TODO: play start audio cue
+					if (!this._decoration) {
+						this._createDecoration();
+					}
+					break;
+				case SpeechToTextStatus.Recognizing: {
+					this._updateInput(e);
```

Comment:
For some reason I do not see the intermediate text when I speak. `SpeechToTextStatus.Recognizing` is firing rapidly with words while you speak and `SpeechToTextStatus.Recognized` when the sentence is done. It would be pretty cool if the results from `Recognizing` appear with a slight grey tone and the result of `Recognized` in normal color when its final.

Reply:
Cool idea, yeah, I think there's something wrong with that state ATM. I was seeing duplicated input when I did get text to show up on `SpeechToTextStatus.Recognizing`.